### ISSUE TITLE
[CSA] False negatives in short bounded loops

### ISSUE THREAD

**Issue by chisa22** [llvm/llvm-project#214227]
Hi, I found a family of false negatives in short bounded loops. Definite bugs at a later iteration, or diagnostics emitted when a function exits after a short bounded loop, disappear with the default configuration and reappear when the same loop is explored more precisely.

## Array-bound reproducer

```c
int data[4];

void write_past_end_in_bounded_loop(void) {
  for (int i = 0; i <= 4; ++i)
    data[i] = i;
}
```

Default analysis emits no warning:

```sh
clang --analyze \
  -Xclang -analyzer-checker=security.ArrayBound \
  bounded_loop_array.c
```

Enabling loop unrolling reports the definite fifth-iteration overflow:

```sh
clang --analyze \
  -Xclang -analyzer-checker=security.ArrayBound \
  -Xclang -analyzer-config -Xclang unroll-loops=true \
  bounded_loop_array.c
```

```text
bounded_loop_array.c:5:5: warning: Out of bound access to memory after the end of 'data'
```

Running the same program with AddressSanitizer confirms a global-buffer-overflow immediately after the 16-byte `data` object.

## Leak reproducer

```cpp
void leak_after_bounded_loop() {
  new int(42);
  for (int i = 0; i < 4; ++i) {
  }
}
```

The default `cplusplus.NewDeleteLeaks` analysis emits no warning. With `unroll-loops=true`, it emits `Potential memory leak` at the loop. Bounds 0 through 3 retain the report under the default configuration; the report disappears starting at bound 4.

The same late-iteration cutoff affects a deterministic division case: a counter initialized to five is decremented to zero before division, but the diagnostic appears only when the initial value is reduced to three or less.

CSA reproduction for the leak case:

```sh
clang++ --analyze \
  -Xclang -analyzer-checker=cplusplus.NewDeleteLeaks \
  bounded_loop_leak.cpp
```

Comparison run:

```sh
clang++ --analyze \
  -Xclang -analyzer-checker=cplusplus.NewDeleteLeaks \
  -Xclang -analyzer-config -Xclang unroll-loops=true \
  bounded_loop_leak.cpp
```

## Comparison with related programs

For the array-bound case, a related reported program reaches the same invalid access before the default loop cutoff:

```c
int data[4];

void write_past_end_in_shorter_loop(void) {
  for (int i = 0; i <= 3; ++i)
    data[i + 1] = i; // warning when i == 3
}
```

The missed program moves the definite invalid access to the later bounded-loop iteration:

```c
int data[4];

void write_past_end_in_bounded_loop(void) {
  for (int i = 0; i <= 4; ++i)
    data[i] = i; // no warning by default; warning with unroll-loops=true
}
```

For the leak case, the direct discarded allocation is reported:

```cpp
void direct_leak() {
  new int(42); // warning
}
```

Adding the short bounded loop after the allocation suppresses the default leak report:

```cpp
void leak_after_bounded_loop() {
  new int(42);
  for (int i = 0; i < 4; ++i) {
  } // no warning by default; warning with unroll-loops=true
}
```

## Expected result

It would be useful for the default analyzer to preserve enough state for small statically bounded loops to diagnose definite bugs reached before or at the loop limit. At minimum, it would help if this default loop-exploration cutoff did not silently suppress unrelated end-of-function leak reports or definite fourth/fifth-iteration errors.

## Verification

CSA version:

```text
clang version 24.0.0git (https://github.com/llvm/llvm-project.git 1cb7e838cd47ecad4050948c0c907ecb1f466ac3)
```

**Comment by steakhal** [comment]
cc @NagyDonat

**Comment by NagyDonat** [comment]
Thanks for the bug report!

We (the analyzer developers) know that loop unrolling mode lets the analyzer report certain kinds of bugs (like the ones that you reported) which are not detected otherwise. A few months ago we evaluated that maybe we should enable loop unrolling by default, but the measurements done by @gamesh411 revealed that it increased the analyzer runtime by +4% [1] and we weren't convinced that the improved coverage justifies this slowdown.

@chisa22 Was this bug report motivated by a situation where the static analyzer failed to find a real bug in a real-world codebase? My intuition was that concretely bounded short (loop unrolling is limited to <= 100 iterations IIRC) loops are rare in real-world code – but if you encountered this "in the wild" then that's motivation to reconsider enabling the loop unrolling by default.

[1] We don't know the exact cause of the slowdown, my personal suspicion is that the "can this loop be unrolled" check is implemented in an inefficient way, but I didn't have time yet to confirm or refute this.
